usr/lib/cca_stdll/cca_specific.c

```diff
@@ -1206,6 +1208,288 @@ CK_RV token_specific_rsa_verify(STDLL_TokData_t * tokdata,
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-224 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA256:
+        if (pss->mgf != CKG_MGF1_SHA256) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-256 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA384:
+        if (pss->mgf != CKG_MGF1_SHA384) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-384 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA512:
+        if (pss->mgf != CKG_MGF1_SHA512) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-512 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    }
+
+    dll_CSNDDSG(&return_code,
+                &reason_code,
+                NULL,
+                NULL,
+                &rule_array_count,
+                rule_array,
+                (long *)&(attr->ulValueLen),
+                attr->pValue,
+                &message_len,
+                message,
+                (long *)out_data_len, &signature_bit_length, out_data);
+
+    if (return_code != CCA_SUCCESS) {
+        TRACE_ERROR("CSNDDSG (RSA PSS SIGN) failed. return :%ld, reason: %ld\n",
+                    return_code, reason_code);
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    } else if (reason_code != 0) {
+        TRACE_WARNING("CSNDDSG (RSA PSS SIGN) succeeded, but "
+                      "returned reason: %ld\n", reason_code);
+    }
+
+done:
+    object_put(tokdata, key_obj, TRUE);
+    key_obj = NULL;
+
+    if (message != NULL)
+        free(message);
+
+    return rc;
+}
+
+CK_RV token_specific_rsa_pss_verify(STDLL_TokData_t *tokdata,
+                                    SESSION  *sess,
+                                    SIGN_VERIFY_CONTEXT *ctx,
+                                    CK_BYTE *in_data,
+                                    CK_ULONG in_data_len,
+                                    CK_BYTE *out_data,
+                                    CK_ULONG out_data_len)
+{
+    CK_RSA_PKCS_PSS_PARAMS *pss;
+    long return_code, reason_code, rule_array_count, message_len;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0, };
+    CK_ATTRIBUTE *attr;
+    OBJECT *key_obj = NULL;
+    CK_BYTE *message = NULL;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+
+    rc = object_mgr_find_in_map1(tokdata, ctx->key, &key_obj, READ_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("object_mgr_find_in_map1 failed\n");
+        goto done;
```

Comment:
sure you want to go down and do a object_put() here ?
wouldn't return rc be a better choice here ?

Reply:
see above